packages/integrations/vercel/src/serverless/middleware.ts

```diff
@@ -61,20 +65,20 @@ function edgeMiddlewareTemplate(middlewarePath: string, vercelEdgeMiddlewareHand
 import { onRequest } from ${middlewarePath};
 import { createContext, trySerializeLocals } from 'astro/middleware';
 export default async function middleware(request, context) {
-	const url = new URL(request.url);
 	const ctx = createContext({
 		request,
 		params: {}
 	});
 	ctx.locals = ${handlerTemplateCall};
-	const next = async () => {
-		const response = await fetch(url, {
+	const { origin } = new URL(request.url);
+	const next = () =>
+		fetch(new URL('/_render', request.url), {
 			headers: {
-				${JSON.stringify(ASTRO_LOCALS_HEADER)}: trySerializeLocals(ctx.locals)
+				...Object.fromEntries(request.headers.entries()),
+				'x-astro-path': request.url.replace(origin, ''),
+				'x-astro-locals': trySerializeLocals(ctx.locals)
 			}
-		});
-		return response;
-	};
+		})
```

Comment:
The edge function calls the node server at `/_render`, with the original path as a header.